### PR TITLE
No longer actively open legacy substreams

### PR DESCRIPTION
**Pull request by tomaka**
Tackles bullet number 2 in [this comment](https://github.com/paritytech/substrate/issues/5670#issuecomment-690117333).

Based on top of https://github.com/paritytech/substrate/pull/7075
Shouldn't be merged now, as we need to publish a version between #7075 and this.
The intention in this PR is to check whether CI is green to make sure that #7075 is working properly. If we merge a broken version of #7075, then we will have to wait again for a bugfix release.

This PR changes `legacy.rs` to no longer pro-actively open a legacy substream.
A consequence of this change is that we can no longer establish outgoing connections to nodes that don't have #7075 (hence the need for a release). However we can still receive incoming connections from nodes that don't have #7075.

The diff is quite large because of all the side clean-ups, but the core part of the changes is that `legacy.rs` no longer emits `OutboundSubstreamRequest`.

I've opted to keep the timeout system on the listening side as long as #7074 isn't resolved. After 60 seconds of inactivity on the legacy substream, the connection is force-closed, thereby freeing the peerset slot.

### CODE REVIEW

**Comment by tomaka** [comment]
Switching to non-draft, as the point is to run the CI.

**Comment by tomaka** [comment]
> I've opted to keep the timeout system on the listening side as long as #7074 isn't resolved. After 60 seconds of inactivity on the legacy substream, the connection is force-closed, thereby freeing the peerset slot.

I've now realized that, since we no longer open legacy substreams, all connections between peers would always close after 60 seconds.

At the moment, the objective of this PR is to prove that https://github.com/paritytech/substrate/pull/7075 works well. I'll fix that after #7075 is approved or merged.

**Comment by mxinden** [comment]
@tomaka let me know once you would like another review on this pull request.

**Comment by tomaka** [comment]
Ready for review.

I've removed the timeout system from the legacy substream entirely.

With notification protocols, the listening side is pro-actively trying to open substreams, which, if they get refused, will result in the keep-alive system closing the connection.

The reason for a timeout system for the legacy substream comes from the situation where the dialer doesn't support notification substreams, and we don't know whether it intends to open a legacy substream.
This is no longer relevant, as nodes that don't support notification protocol substreams are too old to be supported anyway.

**Comment by tomaka** [comment]
Needs a burnin, but after the release of 0.8.24.

File: client/network/src/protocol/generic_proto/handler/legacy.rs [hunk 1]
```diff
@@ -525,24 +478,19 @@ impl ProtocolsHandler for LegacyProtoHandler {
...
 	}
 
 	fn connection_keep_alive(&self) -> KeepAlive {
 		match self.state {
-			ProtocolState::Init { .. } | ProtocolState::Opening { .. } |
-			ProtocolState::Normal { .. } => KeepAlive::Yes,
-			ProtocolState::Disabled { .. } | ProtocolState::Poisoned |
-	  		ProtocolState::KillAsap => KeepAlive::No,
+			ProtocolState::Init { .. } | ProtocolState::Normal { .. } => KeepAlive::Yes,
+			ProtocolState::Opening { .. } | ProtocolState::Disabled { .. } |
```
Comment (tomaka) [inline client/network/src/protocol/generic_proto/handler/legacy.rs:492]: `Opening` is now `No` because of the removal of the timeout.

File: client/network/src/protocol/generic_proto/handler/group.rs [hunk 2]
```diff
@@ -488,45 +474,30 @@ impl ProtocolsHandler for NotifsHandler {
 
 	fn inject_dial_upgrade_error(
 		&mut self,
-		num: Option<usize>,
-		err: ProtocolsHandlerUpgrErr<EitherError<NotificationsHandshakeError, io::Error>>
+		num: usize,
+		err: ProtocolsHandlerUpgrErr<NotificationsHandshakeError>
 	) {
 		match (err, num) {
```
Comment (mxinden) [inline client/network/src/protocol/generic_proto/handler/group.rs:480]: I don't think it still makes sense to match on `num`, right?

**Review by mxinden — APPROVED** [review]
> Needs a burnin, but after the release of 0.8.24.

:+1:

**Review by romanb — APPROVED** [review]

**Comment by tomaka** [comment]
Let's wait until Monday to start the burnin, so that more nodes have upgraded.

**Comment by tomaka** [comment]
Burnin' report:
- The percentage of filled slots seems to be unable to go above 70%.
- The number of connections forcibly closed by the local node has dropped to almost 0.
- The number of connections forcibly closed by the remote has almost doubled.
- The number of connections closed because of the keep alive timeout increased from ~5/mn to ~20/mn
- The number of dialing errors caused by an invalid PeerId is skyrocketting, from ~25/mn to ~200/mn

Force-closing a connection is almost always caused by the dialing side not opening a legacy substream in time when the listening side has reserved a slot (see #7074). As expected, the node with this PR consequently doesn't force-close connections anymore.
Instead, the situation where the dialer doesn't open substreams when expected is now supposed to trigger the keep-alive timeout rather than force-closing the connection.
Since the node with this PR no longer opens a legacy substream, its outgoing connections are force-closed by nodes on 0.8.23 and below.

In other words, so far the observation is consistent with what is expected. It's disappointing that only ~20% of the network (roughly guessing from looking at the telemetry) seems to be using 0.8.24, which is not enough to even fill all the slots of a single node.

**Comment by tomaka** [comment]
> The percentage of filled slots seems to be unable to go above 70%. While the metrics don't distinguish between in and out slots, my assumption (as expanded in details below) is that all in slots are full while out slots aren't.

Filled slots now at 85%, which probably follows the nodes upgrading to 0.8.24.

**Comment by tomaka** [comment]
The node looks like it is behaving normally.
We only worry concerns the increase in the `InvalidPeerId` errors. See also #7198.

**Comment by tomaka** [comment]
I've been informed that the increase in `InvalidPeerId` errors corresponds to moment when the node had `--reserved-nodes` flags. For 2 days, the PR got burned-in without these flags, and it looked normal. I don't think in general that this `InvalidPeerId` problem is related in any way to this PR, and would go for merging.

**Comment by mxinden** [comment]
> and would go for merging.

As far as I understand this pull request requires https://github.com/paritytech/substrate/pull/7075 to be widely deployed. https://github.com/paritytech/substrate/pull/7075 is only part of Polkadot v0.8.24. When we merge this pull-request now it will be part of Polkadot v0.8.25.

Are we sure right now that v0.8.24 will be widely deployed once v0.8.25 is released?

**Comment by tomaka** [comment]
We had a couple of announcements asking validators to upgrade to at least 0.8.24.
One can see on telemetry that approximately 1/3rd of the network still uses 0.8.23-, which in the absolute is still high but low enough that merging this PR isn't really risky anymore.

**Comment by tomaka** [comment]
bot merge
